src/main/java/br/com/caelum/viagens/voos/utils/GrafoRotasUtils.java

```diff
@@ -0,0 +1,95 @@
+package br.com.caelum.viagens.voos.utils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.shortestpath.AllDirectedPaths;
+import org.jgrapht.graph.DefaultDirectedGraph;
+
+import br.com.caelum.viagens.administrativo.model.Aeroporto;
+import br.com.caelum.viagens.support.Route;
+
+public class GrafoRotasUtils {
```

Comment:
*Avaliamos de forma lógica o nível de complicação de cada trecho de código. Aqui não tem espaço para feeling. Usamos a teoria da carga cognitiva e medimos o nível de pontos de entendimentos necessário por trecho.*

Essa aqui você me explica :). Muito legal que você experimentou usar o lance do grafo!

Reply:
eu vou tentar explicar aqui como esta funcionando, rs 

primeiro eu crio o grafo no método grafoDeRotas() - para cada rota serão criados vértices para origem e destino com a ligação entre eles (a aresta) sendo a própria rota. eu aproveitei a classe AllDirectPaths da biblioteca JGraphT que vai montar todos os caminhos possíveis do grafo, ou seja, todos os caminhos possíveis entre os vertices (aeroportos) - que é feito no método getAllCaminhosDiretos().

a classe AllDirectPaths possui o método getAllPaths() que me devolve todos os caminhos possíveis do grafo especificando uma coleção de origens e destinos. no método getCaminhos() que recebe apenas as rotas, eu uso o getAllPaths() para pegar todos os caminhos possíveis entre todas as origens e todos os destinos (que são candidatos a perna final) de cada rota. aqui, "candidatos a perna final" são todos os destinos que não são origem de outra rota.

feito isso, no método temSequenciaLogica() eu verifico se existe um caminho que possui o mesmo tamanho que minha lista de rotas original. se for verdadeiro, isso quer dizer que as rotas possuem sequencia lógica e foi encontrado um caminho direto. se falso, o contrário, já que até podem existir caminhos, mas eles podem ser menores (caso uma rota esteja desconexa das demais) ou pode ser maior (com rotas se repetindo e aumento o percurso).

também criei o método getCaminhos() que além de receber as rotas, recebe os aeroportos de origem inicial e destino final do voo para facilitar esta validação específica que eu utilizo para validar no construtor da classe Voo. além de métodos para identificar a origem inicial e destino final das rotas para montar a lista em sequencia lógica para ser devolvida para o cliente.

eu não estudei a fundo os algoritmos que a biblioteca disponibiliza, talvez tenha um que facilite ainda mais o código.

Reply:
Chique! Aqui eu preciso ler umas 3x :).